pubsublite: Add OpenTelemetry instrumentation

dmathieu opened the pull request:
This adds instrumentation for the pubsublite consumer.

marclop commented on pubsublite/internal/telemetry/publisher.go:
```diff
@@ -0,0 +1,66 @@
+
+	res := h(ctx, msg)
+
+	go func() {
+		mid, err := res.Get(ctx)
```
I don't like a couple of things about this approach:

- `res.Get()` is called twice, here and in `pubsublite/producer.go`.
- A goroutine is created for each published message. This may create quite a lot of overhead in the Go runtime for mid-high throughput producers.

I don't have a great alternative for this right now, but maybe worth dropping a NOTE in the code?

marclop replied:
I also came across: https://github.com/googleapis/google-cloud-go/issues/2953. Perhaps we can add this link to the comment so we've a bit more context?

Would you mind opening an issue in this repo about the potential problems of this approach? One way to possibly mitigate this would be to return the span in the Result, and handle the end of the span lifecycle as a callback function. We're already handling the responses back in the `pubsublite/producer.go`, so calling an extra function when the response is obtained may not be a bad thing to do. The referenced issue has some details on how messages are batched and sent too.

marclop reviewed (commented):
Looks pretty good! I'm concerned about some potential goroutine explosion for tracing and what that's going to mean for applications that have instrumentation turned on.

dmathieu commented on pubsublite/internal/telemetry/publisher.go:
```diff
@@ -0,0 +1,66 @@
+		),
+	)
+
+	if msg == nil {
+		msg = &pubsub.Message{}
+	}
```
Yeah, it shouldn't happen. It was more to be thorough in tests.

kruskall commented on pubsublite/internal/telemetry/consumer_test.go:
```diff
@@ -0,0 +1,171 @@
+				spans[i].EndTime = time.Time{}
+			}
+
+			assert.Equal(t, spans, tt.expectedSpans)
```
```suggestion
			assert.Equal(t, tt.expectedSpans, spans)
```

kruskall commented on pubsublite/internal/telemetry/consumer_test.go:
```diff
@@ -0,0 +1,171 @@
+
+func TestConsumer(t *testing.T) {
+	exp := tracetest.NewInMemoryExporter()
+	tp := sdktrace.NewTracerProvider(
```
**suggestion**: we should shutdown the tracer to release any resource

kruskall commented on pubsublite/internal/telemetry/consumer_test.go:
```diff
@@ -0,0 +1,171 @@
+	var err error
+	t, err = trace.TraceIDFromHex(s)
+	if err != nil {
+		panic(err)
```
**suggestion**: should we use testify `require.NoError` for consistency ?

kruskall commented on pubsublite/internal/telemetry/consumer_test.go:
```diff
@@ -0,0 +1,171 @@
+	}
+}
+
+func mustTraceIDFromHex(s string) (t trace.TraceID) {
```
issue: we need a `t.Helper` here to print the correct line number in case of failure

kruskall commented on pubsublite/internal/telemetry/publisher_test.go:
```diff
@@ -0,0 +1,160 @@
+
+func TestPublisher(t *testing.T) {
+	exp := tracetest.NewInMemoryExporter()
+	tp := sdktrace.NewTracerProvider(
```
**issue**: same as above, we should shutdown the tracer

kruskall commented on pubsublite/internal/telemetry/publisher_test.go:
```diff
@@ -0,0 +1,160 @@
+				}
+			}
+
+			assert.Equal(t, spans, tt.expectedSpans)
```
```suggestion
			assert.Equal(t, tt.expectedSpans, spans)
```

dmathieu commented on pubsublite/internal/telemetry/consumer.go:
```diff
@@ -0,0 +1,60 @@
+type consumerHandler = func(context.Context, *pubsub.Message)
+
+// Consumer adds telemetry data to messages received
+func Consumer(tracer trace.Tracer, h consumerHandler) consumerHandler {
```
I'm adding those attributes.
I've chosen to add an attributes parameter, even though it feels a bit weird to expose otel to `consumer.go`. I could also add the subscription to the consumer struct, so we get the additional information within the tracer.

dmathieu commented on pubsublite/internal/telemetry/publisher.go:
```diff
@@ -0,0 +1,66 @@
+type producerHandler = func(context.Context, *pubsub.Message) *pubsub.PublishResult
+
+// Publisher adds telemetry data to messages published
+func Publisher(ctx context.Context, tracer trace.Tracer, msg *pubsub.Message, h producerHandler) *pubsub.PublishResult {
```
As attributes, or elsewhere? I've added name, project and region as attributes.

marclop replied:
Yes as an attribute, [`semconv.MessagingDestinationNameKey`](https://pkg.go.dev/go.opentelemetry.io/otel@v1.14.0/semconv/v1.18.0#MessagingDestinationNameKey) seems appropriate. Where have you added the region and project in the publisher? I see you added them in the consumer, but not the publisher.

dmathieu replied:
Ah, yes, sorry. I had missed the publisher. I just added it.

dmathieu commented on pubsublite/internal/telemetry/publisher_test.go:
```diff
@@ -0,0 +1,160 @@
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			res := &pubsub.PublishResult{}
```
The thing is that it's a channel we have to close. And `reflect` will not us call [Close](https://pkg.go.dev/reflect#Value.Close) on a channel retrieved from a private value 😢

marclop replied:
Ah, I missed that, well then, not much we can do

dmathieu replied:
I've opened an issue: https://github.com/googleapis/google-cloud-go/issues/7801

marclop replied:
Thanks!

marclop commented on pubsublite/internal/telemetry/publisher.go:
```diff
@@ -0,0 +1,68 @@
+	ctx, span := tracer.Start(ctx, "pubsublite.Publish",
+		trace.WithSpanKind(trace.SpanKindProducer),
+		trace.WithAttributes(
+			semconv.MessagingSystemKey.String("pubsub"),
```
```suggestion
			semconv.MessagingSystemKey.String("pubsublite"),
```

marclop commented on pubsublite/consumer.go:
```diff
@@ -171,6 +175,11 @@ func NewConsumer(ctx context.Context, cfg ConsumerConfig) (*Consumer, error) {
 			),
+			telemetryAttributes: []attribute.KeyValue{
+				semconv.MessagingSourceNameKey.String(subscription.Name),
+				attribute.String("region", subscription.Region),
```
Shouldn't this be using [`semconv.CloudRegion`](https://pkg.go.dev/go.opentelemetry.io/otel@v1.14.0/semconv/v1.18.0#CloudRegion)? 

```suggestion
				semconv.CloudRegion(subscription.Region),
```

marclop commented on pubsublite/consumer.go:
```diff
@@ -171,6 +175,11 @@ func NewConsumer(ctx context.Context, cfg ConsumerConfig) (*Consumer, error) {
+			telemetryAttributes: []attribute.KeyValue{
+				semconv.MessagingSourceNameKey.String(subscription.Name),
+				attribute.String("region", subscription.Region),
+				attribute.String("project", subscription.Project),
```
```suggestion
				semconv.CloudAccountID(subscription.Project),
```

kruskall reviewed (approved):
LGTM!

marclop reviewed (commented):
The kafka and this implementation seem to differ slightly in what can be configured:

In Kafka we've got a `TraceProvider trace.TracerProvider` and `DisableTelemetry bool`. Wouldn't it be better if both packages accepted the same flags to enable and configure the instrumentation?

dmathieu commented:
I've added a `TracerProvider` config.
For `DisableTelemetry`, I'm not sure. I added it in case the hook had unintended side effects. In the case of manual telemetry though, we own the code and can disable telemetry globally by not configuring an SDK.

Similar discussion: https://github.com/elastic/apm-queue/pull/104#discussion_r1173292307
